test/src/main/java/org/apache/accumulo/test/functional/HalfClosedTabletIT.java

```diff
@@ -0,0 +1,219 @@
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import com.google.common.collect.Sets;
+
+// This IT tests the cases seen in https://github.com/apache/accumulo/issues/3674
+// where a failing minor compaction causes a Tablet.initiateClose to leave the
+// Tablet in a half-closed state. The half-closed Tablet cannot be unloaded and
+// the TabletServer cannot be shutdown normally. Because the minor compaction has
+// been failing the Tablet needs to be recovered when it's ultimately re-assigned.
+public class HalfClosedTabletIT extends SharedMiniClusterBase {
+
+  public static class HalfClosedTabletITConfiguration implements MiniClusterConfigurationCallback {
+
+    @Override
+    public void configureMiniCluster(MiniAccumuloConfigImpl cfg, Configuration coreSite) {
+      cfg.setNumTservers(1);
+    }
+
+  }
+
+  @BeforeAll
+  public static void startup() throws Exception {
+    SharedMiniClusterBase.startMiniClusterWithConfig(new HalfClosedTabletITConfiguration());
+  }
+
+  @AfterAll
+  public static void shutdown() throws Exception {
+    SharedMiniClusterBase.stopMiniCluster();
+  }
+
+  @Test
+  public void testSplitWithInvalidContext() throws Exception {
+
+    // In this scenario the table has been mis-configured with an invalid context name.
+    // The minor compaction task is failing because classes like the volume chooser or
+    // user iterators cannot be loaded. The user calls Tablet.split which calls initiateClose.
+    // This test ensures that the Tablet can still be unloaded normally by taking if offline
+    // after the split call with an invalid context. The context property is removed after the
+    // split call below to get the minor compaction task to succeed on a subsequent run. Because
+    // the minor compaction task backs off when retrying, this could take some time.
+
+    String tableName = getUniqueNames(1)[0];
+
+    try (final var client = Accumulo.newClient().from(getClientProps()).build()) {
+      final var tops = client.tableOperations();
+      tops.create(tableName);
+      TableId tableId = TableId.of(tops.tableIdMap().get(tableName));
+      try (final var bw = client.createBatchWriter(tableName)) {
+        final var m1 = new Mutation("a");
+        final var m2 = new Mutation("b");
+        m1.put(new Text("cf"), new Text(), new Value());
+        m2.put(new Text("cf"), new Text(), new Value());
+        bw.addMutation(m1);
+        bw.addMutation(m2);
+      }
+
+      setInvalidClassLoaderContextPropertyWithoutValidation(getCluster().getServerContext(),
+          tableId);
+
+      Thread.sleep(500);
+
+      // This should fail to split, but not leave the tablets in a state where they can't
+      // be unloaded
+      assertThrows(AccumuloServerException.class,
+          () -> tops.addSplits(tableName, Sets.newTreeSet(List.of(new Text("b")))));
+
+      removeInvalidClassLoaderContextPropertyWithoutValidation(getCluster().getServerContext(),
+          tableId);
```

Comment:
I tried removing the code in initiateClose and running the test locally.  Made the following changes to this test to get it running.

```suggestion
      Thread configFixer = new Thread(() -> {
        UtilWaitThread.sleep(3000);
        removeInvalidClassLoaderContextPropertyWithoutValidation(getCluster().getServerContext(),
                tableId);
      });

      // grab this time before starting the thread starts and before it sleeps.
      long t1 = System.nanoTime();
      configFixer.start();

      // The split will probably start running w/ bad config that will cause it to get stuck.  However once the config is fixed by the background thread it should continue.
      tops.addSplits(tableName, Sets.newTreeSet(List.of(new Text("b"))));
      long t2 = System.nanoTime();

      // expect that split took at least 3 seconds because that is the time it takes to fix the config
      assertTrue(TimeUnit.NANOSECONDS.toMillis(t2-t1) > 3000);
```

Reply:
My guess is that it works because I changed the catch clause in MinorCompactionTask from IOException to Exception. If you were to revert that single change, then I think this would fail and leave the tablet in a bad state.

Reply:
I made these changes in 9f84ed4, however 2/3 of these new tests no longer work. I believe that is because I merged 3678 into 2.1 and into this branch. The tests that are failing are setting an invalid context, and when ClassLoaderUtil.getClassLoader is called, the context is not valid, and the non-context aware classloader is returned instead of the code returning null or throwing an error.

Reply:
@keith-turner - If #3683 is merged, then I think that these tests will work again.